core/src/main/java/org/opensearch/sql/planner/physical/LookupOperator.java

```diff
@@ -0,0 +1,131 @@
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.ToString;
+import org.opensearch.sql.data.model.ExprTupleValue;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.data.model.ExprValueUtils;
+import org.opensearch.sql.expression.ReferenceExpression;
+
+/** Lookup operator. Perform lookup on another OpenSearch index and enrich the results. */
+@Getter
+@ToString
+@EqualsAndHashCode(callSuper = false)
+public class LookupOperator extends PhysicalPlan {
+  @Getter private final PhysicalPlan input;
+  @Getter private final String indexName;
+  @Getter private final Map<ReferenceExpression, ReferenceExpression> matchFieldMap;
+  @Getter private final Map<ReferenceExpression, ReferenceExpression> copyFieldMap;
+  @Getter private final Boolean appendOnly;
+
+  @EqualsAndHashCode.Exclude
+  private final BiFunction<String, Map<String, Object>, Map<String, Object>> lookup;
+
+  /** Lookup Constructor. */
+  @NonNull
+  public LookupOperator(
+      PhysicalPlan input,
+      String indexName,
+      Map<ReferenceExpression, ReferenceExpression> matchFieldMap,
+      Boolean appendOnly,
+      Map<ReferenceExpression, ReferenceExpression> copyFieldMap,
+      BiFunction<String, Map<String, Object>, Map<String, Object>> lookup) {
+    this.input = input;
+    this.indexName = indexName;
+    this.matchFieldMap = matchFieldMap;
+    this.appendOnly = appendOnly;
+    this.copyFieldMap = copyFieldMap;
+    this.lookup = lookup;
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitLookup(this, context);
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  @Override
+  public boolean hasNext() {
+    return input.hasNext();
+  }
+
+  @Override
+  public ExprValue next() {
+    ExprValue inputValue = input.next();
+
+    if (STRUCT == inputValue.type()) {
+      Map<String, Object> matchMap = new HashMap<>();
+      Map<String, Object> finalMap = new HashMap<>();
+
+      for (Map.Entry<ReferenceExpression, ReferenceExpression> matchField :
+          matchFieldMap.entrySet()) {
+        Object val = inputValue.bindingTuples().resolve(matchField.getValue()).value();
+        if (val != null) {
+          matchMap.put(matchField.getKey().toString(), val);
+        } else {
+          // No value in search results, so we just return the input
+          return inputValue;
+        }
+      }
+
+      finalMap.put("_match", matchMap);
+
+      Map<String, String> copyMap = new HashMap<>();
+
+      if (!copyFieldMap.isEmpty()) {
+
+        for (Map.Entry<ReferenceExpression, ReferenceExpression> copyField :
+            copyFieldMap.entrySet()) {
+          copyMap.put(String.valueOf(copyField.getKey()), String.valueOf(copyField.getValue()));
+        }
+
+        finalMap.put("_copy", copyMap.keySet());
+      }
+
+      Map<String, Object> lookupResult = lookup.apply(indexName, finalMap);
```

Comment:
The complexity is O(N), N is rows in input dataset?  should we rewrite query as LEFT JOIN? 
Do we have performance test result?

Reply:
No peformance tests yet.
Running the lookup command as a join has also perf related implications like triggering the circuit breaker when both indices contains lot of documents and fields. If we have perf related concerns maybe keeping the current logic and adding a simple LRU cache is an option? And how would one implement then "appendonly" flag with a join?